### ISSUE TITLE
Document synchronization semantics of View initialization

### ISSUE THREAD

**Issue by PhilMiller** [kokkos/kokkos#4697]
what execution space / stream should it run on, and how much should it synchronize with besides the caller?

```c++
View<double*> a("A",N); 
// which stream does this go into, is the data zeroed out by the time the call returns
// does this imply Kokkos::fence() or just execution_space().fence() or not even that?
View<double*> b(view_alloc(WithoutInitializing,"B"),N);
View<double*> c(view_alloc(exec,"C"),N);
```
* what if we had asynchronous allocation
   * that would make the view non-copyable (i.e. b = c) before synching which is shitty
* normal allocations not syncing at all could be problem for UVM / or async host backends

```c++
    View<double*,HostSpace> a("A",N);
    a(2) = 5; // OOPS with HPX
```

**Comment by PhilMiller** [comment]
For an application that's conscientiously using multiple streams, the potential for case `A` to imply a global `Kokkos::fence()` would have catastrophic performance implications if any code naively allocates a View during otherwise asynchronous execution

**Comment by PhilMiller** [comment]
That 'naive' code may appear inside a library function that the application calls but doesn't control, which makes this a substantial composability issue.

**Comment by masterleinad** [comment]
That's pretty much the exact same issue as with the two-argument vs. three-argument deep copy, though, right? (Just that we are much clearer about the semantics for that one already).
We rely on library functions to do the right thing and only fence the stream that we are operating on.

**Comment by PhilMiller** [comment]
Alternatives to `Kokkos::fence()` would be `execution_space().fence()` as noted, or 

```c++
auto instances = Kokkos::Experimental::partition_space(execution_space(), 1);
deep_copy(instances[0], new_view, 0);
instances[0].fence();
```

**Comment by PhilMiller** [comment]
Yes, it's the same distinction as between two and three argument `deep_copy`, just that it's a lot easier to systematically modify and/or detect the undesirable `deep_copy` cases than the View initializations with arbitrary types and unspecified underlying operations

**Comment by PhilMiller** [comment]
Another alternative would be a dedicated internal execution space instance used for all initializations, and running and fencing specifically on that. Abstractly, the fresh stream is preferable, but if stream creation itself is expensive, then this would be a good fallback.

**Comment by PhilMiller** [comment]
If we do without an internal fence entirely in case A, the caller would either need to call `Kokkos::fence()` to synchronize when it's eventually needed, or have a new explicit interface to synchronize however the initialization was launched.

**Comment by PhilMiller** [comment]
The problem is slightly stickier for views of classes with constructors that set up meaningful state. Hypothetically, either the view constructor must run a kernel to construct them (a la `std::vector`), or if using `WithoutInitializing`, the client code must do something like

```c++
class MyClass {};
View<MyClass*> v("mine", N);
parallel_for(N, KOKKOS_LAMBDA(int i) {
new (&v(i)) MyClass(/*whatever args*/);
});
```

Since even `MyClass::operator=` invoked by `v(i) = MyClass();` would technically be undefined behavior, since the lifetime of the view element would not have started, but a class method is being invoked on it.

I feel that this leans pretty hard toward providing some sort of guaranteed complete initialization before the case-A `View` constructor returns to the caller.

**Comment by PhilMiller** [comment]
Another alternative design, which I don't particularly endorse but that we may find reasons to use, is to have distinct `thread_local` execution space instances to initialize and fence on.

**Comment by mhoemmen** [comment]
Following since I have an interest.  My personal view is that Views should be inaccessible except inside an "asynchronous parallel region" with a clearly defined execution space, but it's a bit too late to close that barn door ; - ) .

**Comment by PhilMiller** [comment]
@ajpowelsnl Yes, this is on deck for further discussion. I'll summarize the present state of things here.

@mhoemmen That's totally a possibility for 4.0. Not likely to fly with anyone, but totally a possibility!

The note about asynchronous allocation being off the table because of the impossibility of copying (and such) may be too conservative. The pointer to the underlying memory in whatever space lives in the shared record that Kokkos allocates in the heap in host memory. The individual `View` instances point at that, and that pointer is perfectly fine to copy before the record has the `data()` pointer filled in. They'd just be of very limited use, since any operation that does need `data()` would be forbidden until synchronization, or unless sequenced in the stream that's doing the allocation. There may also be something about how `View`s are captured for use in `parallel_for` et al that requires the backing record to be fully initialized beforehand.

**Comment by PhilMiller** [comment]
There's a range of synchronization possibilities to consider

1 Global Synchronization
```c++
allocate();
parallel_for(initialize);
Kokkos::fence();
```

2 Default stream used for all general allocation and initialization
```c++
auto e = execution_space();
allocate();
parallel_for(e, initialize);
e.fence();
```

3 Dedicated stream - global
```c++
 // static storage duration, not local/temporary
static auto e = partition_space(execution_space());

allocate();
parallel_for(e, initialize);
e.fence();
```

4 Dedicated stream - per host thread
```c++
 // static storage duration, not local/temporary
thread_local auto e = partition_space(execution_space());

allocate();
parallel_for(e, initialize);
e.fence();
```

5 One-off stream
```c++
auto e = partition_space(execution_space());
allocate();
parallel_for(e, initialize);
e.fence();
```

Q Fully asynchronous allocation and initialization, with no implied fence, and an added method on `View` to let callers fence as needed (or stick in a graph, or whatever)

Z Don't fence, maybe specify what execution space is used, let applications figure it out

**Comment by PhilMiller** [comment]
From what I understand of the way CUDA schedules execution, with overall FIFO behavior, even 5 may not ultimately return all that much sooner than 1. This may need testing, and may push toward the extreme later options.

**Comment by mhoemmen** [comment]
@PhilMiller wrote:
> That's totally a possibility for 4.0. Not likely to fly with anyone, but totally a possibility!

I know y'all won't win that fight, alas.  Users don't bother building abstractions on their own, and they complain if they can't write the same stateful mutable classes that were already old-fashioned 20 years ago.  Also they hate callbacks for some reason I don't fathom.

**Comment by PhilMiller** [comment]
Given Christian's comments about interesting behavior of the HPX backend even with `HostSpace` allocations, perhaps @msimberg could comment on this?

**Comment by msimberg** [comment]
For HPX the ideal situation would be something like Q, i.e. the view constructor takes an explicit execution space and does not fence at all. I suppose that's also sort of what @mhoemmen hopes to have since that requires either an explicit fence before access by the user or accesses only within parallel regions (on the correct instance): 

> My personal view is that Views should be inaccessible except inside an "asynchronous parallel region" with a clearly defined execution space

When using the HPX backend ironically fencing the HPX backend itself is not so bad since `HPX::fence()` plays nicely with HPX threads, i.e. yields the HPX thread if needed. The problem is fencing on other backends that would actually block the HPX thread. If fencing is left to the user it means we can hide the synchronization behind a thin layer as we do in https://github.com/STEllAR-GROUP/hpx-kokkos. E.g. `parallel_for_async(cuda_instance, ...)` returns a future which is synchronized on the stream of `cuda_instance` with HPX's internal synchronization mechanisms. Likewise a non-fencing view constructor could then be hidden by us behind a `future<View> create_view_async(...)` which does synchronize correctly. Currently https://github.com/STEllAR-GROUP/octotiger assumes that the view constructor is blocking and because of this sets up a pool of unmanaged views at the beginning of the application to avoid unwanted fences.

TL;DR: No fences gives HPX the freedom to do synchronization with it's own mechanisms (but I can see how that may be confusing if one is using "plain Kokkos", not a "futurized Kokkos").

**Comment by PhilMiller** [comment]
@msimberg That kind of wrapping layer to control synchronization is exactly what I've been working on in DARMA/vt and EMPIRE, (down to a comparable `parallel_for_async` routine) and is what led me to raise this discussion. We want to be able to trust that calls to library code won't naively screw our careful efforts by inadvertently fencing more than they should/need.

**Comment by crtrott** [comment]
(The following may break any code even ones which never create explicit instances)
I feel good about using a dedicated (internal instance) to initialize Views for the standard construction (`View<..> a("A",N);`). View creation would potentially not block on already launched work in any instance. This would potentially introduce a bug where code implicitly relied on this fence.

SF F N A SA
2 4 5 5 0 

(The following may break code which uses explicit instances)
I feel good about using the DEFAULT instance to initialize Views for the standard construction (`View<..> a("A",N);`). View creation would potentially not block on already launched work in other instances. This would potentially introduce a bug where code implicitly relied on this fence when using other explicit instances.

SF F N A SA
2 8 6 1 0

**Comment by PhilMiller** [comment]
Ok, so I guess the first order of business will  be writing a test using the `kokkosp` interfaces to check the desired behavior - an initialization kernel on the default instance followed by a fence on the default instance, and no global fence.
